src/corehost/common/trace.h

```diff
@@ -31,6 +33,39 @@ namespace trace
 
     // Returns the currently set callback for error writing
     error_writer_fn get_error_writer();
+
+    void trace(const pal::char_t* format, ...);
+    void trace_error(const pal::char_t* format, ...);
+
+    void println(const pal::char_t* format, ...);
+    void println();
+    void flush();
 };
 
+// Tracing is performed with a macro so that arguments to trace::trace() and
+// trace::trace_error() can be evaluated lazily depending on the current
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
+    do {                                                            \
+        if (Verbosity == trace::verbosity::Error)                   \
```

Comment:
I personally would prefer to handle this difference in the macros below - basically have `TRACE_ERROR` implement itself fully and then have the others use the `TRACE_LAZY`.

Reply:
Done.